### PR TITLE
Add DockerContainersEnumerationTask analysis task

### PR DESCRIPTION
**Pull request by rgayon**

### CODE REVIEW

File: turbinia/workers/docker.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+    Returns:
+      a list(dict) containing information about the containers found.
+
+    Raises:
+      TurbiniaException: when the docker-explorer tool failed to run.
+    """
+
+    mount_path = evidence.local_path
+    if type(evidence).__name__ == 'RawDisk':
```
Comment (aarontp) [inline turbinia/workers/docker.py]: ??  Why is this block here?  This should be done in the preprocessor, right?

File: turbinia/workers/docker.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      mount_path = mount_local.PreprocessMountDisk(
+          evidence.loopdevice_path, evidence.mount_partition)
+
+    docker_dir = os.path.join(mount_path, 'var', 'lib', 'docker')
```
Comment (aarontp) [inline turbinia/workers/docker.py]: See above comment about not hard-coding this path.

File: turbinia/workers/docker.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    containers_info = None
+    docker_explorer_command = [
+        'sudo', '/usr/local/bin/de.py', '-r', docker_dir, 'list',
+        'all_containers'
+    ]
+    try:
+      log.info('Running {0:s}'.format(' '.join(docker_explorer_command)))
+      json_string = subprocess.check_output(docker_explorer_command)
```
Comment (aarontp) [inline turbinia/workers/docker.py]: Any reason not to run this through `self.execute()`?

File: turbinia/workers/docker.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    except Exception as e:
+      mount_local.PostprocessUnmountPath(mount_path)
+      raise TurbiniaException(
```
Comment (aarontp) [inline turbinia/workers/docker.py]: If you're going to fail early from this function, we should trap the exception in `run()` and then return the TurbiniaResult with the updated errors/status set.

File: turbinia/workers/docker.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+          'Failed to run {0:s} {1!s}'.format(
+              ' '.join(docker_explorer_command), e))
+
+    try:
+      containers_info = json.loads(json_string)
+    except ValueError as e:
+      mount_local.PostprocessUnmountPath(mount_path)
```
Comment (aarontp) [inline turbinia/workers/docker.py]: Here are a couple more examples of post-processors running in the `run()` method instead of being called implicitly\automatically by the run wrapper.  Same as above, but I'm not sure I understand why that's being done here.

File: turbinia/workers/docker.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      TurbiniaTaskResult object.
+    """
+
+    containers_info = self.GetContainers(evidence)
+    for container_info in containers_info:
+      container_id = container_info.get('container_id')
+      container_evidence = DockerContainer(container_id=container_id)
+      result.add_evidence(container_evidence, evidence.config)
+
+    result.close(self, success=True)
```
Comment (aarontp) [inline turbinia/workers/docker.py]: Maybe set a useful status here too, like "Found N Docker Containers"?

File: turbinia/evidence.py [hunk 7]
```diff
@@ -355,3 +356,44 @@ def __init__(self, artifact_name):
...
+
+  Attributes:
+    container_id(str): The ID of the container to mount.
+  """
+
+  def __init__(self, container_id=None, *args, **kwargs):
+    """Initialization for Docker Container."""
+    super(DockerContainer, self).__init__(*args, **kwargs)
+    self.container_id = container_id
+    self._container_fs_path = None
```
[the remark this answers is not in the document]
Comment (rgayon) [reply turbinia/evidence.py]: Usually (ie; in plaso), when an attribute name starts with _, it's 'private' and shouldn't be exposed in the documentation.
See self._data in turbinia/lib/libcloudforensics.py

File: turbinia/workers/docker.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+    mount_path = evidence.local_path
+    if type(evidence).__name__ == 'RawDisk':
+      mount_path = mount_local.PreprocessMountDisk(
+          evidence.loopdevice_path, evidence.mount_partition)
+
+    docker_dir = os.path.join(mount_path, 'var', 'lib', 'docker')
+    containers_info = None
+    docker_explorer_command = [
+        'sudo', '/usr/local/bin/de.py', '-r', docker_dir, 'list',
```
[the remark this answers is not in the document]
Comment (rgayon) [reply turbinia/workers/docker.py]: this is now in gift PPA where canonical path is /usr/bin.

There is also a packaged module, and we could make turbinia depend on it (instead of raising if we can't find it) WDYT?

File: turbinia/evidence.py [hunk 9]
```diff
@@ -355,3 +356,44 @@ def __init__(self, artifact_name):
...
+    self._docker_root_directory = None
+    self._mount_path = None
+    self.context_dependent = True
+
+  def _preprocess(self):
+    self._mount_path = self.parent_evidence.local_path
```
[the remark this answers is not in the document]
Comment (rgayon) [reply turbinia/evidence.py]: Though this can't happen if self.context_dependent is True, when instantiating a new DockerContainer evidence, I guess I can raise and explain why 'this shouldn't happen'

File: turbinia/workers/docker.py (hunk 6 above)
Comment (rgayon) [reply turbinia/workers/docker.py]: Ack

File: turbinia/workers/docker.py (hunk 4 above)
Comment (rgayon) [reply turbinia/workers/docker.py]: done

File: turbinia/evidence.py [hunk 10]
```diff
@@ -355,3 +356,44 @@ def __init__(self, artifact_name):
...
+    # Mounting the container's filesystem
+    self._container_fs_path = docker.PreprocessMountDockerFS(
+        self._docker_root_directory, self.container_id)
+    self.local_path = self._container_fs_path
+
+  def _postprocess(self):
+    # Unmount the container's filesystem
+    mount_local.PostprocessUnmountPath(self.local_path)
+    self._container_fs_path = None
+    if type(self.parent_evidence).__name__ == 'RawDisk':
```
[the remark this answers is not in the document]
Comment (rgayon) [reply turbinia/evidence.py]: No, this is a specific step that I need to do, because RawDisk pre/postprocessor doesn't mount/unmount the underlying filesystem. (see comment below)

File: turbinia/evidence.py [hunk 11]
```diff
@@ -355,3 +356,44 @@ def __init__(self, artifact_name):
...
+    super(DockerContainer, self).__init__(*args, **kwargs)
+    self.container_id = container_id
+    self._container_fs_path = None
+    self._docker_root_directory = None
+    self._mount_path = None
+    self.context_dependent = True
+
+  def _preprocess(self):
+    self._mount_path = self.parent_evidence.local_path
+    if issubclass(type(self.parent_evidence), RawDisk):
```
[the remark this answers is not in the document]
Comment (rgayon) [reply turbinia/evidence.py]: TODO'ed
Though I think there may be a case where, for example, plaso will be able to do things on a disk, while we won't be able to mount its partitions (ie; 'dirty' FS that the kernel won't know how to mount, etc)

File: turbinia/evidence.py [hunk 12]
```diff
@@ -355,3 +356,44 @@ def __init__(self, artifact_name):
...
+      # Mounting the filesystem on the disk, as this is not done by the RawDisk
+      # evidence preprocessor
+      self._mount_path = mount_local.PreprocessMountDisk(
+          self.parent_evidence.loopdevice_path,
+          self.parent_evidence.mount_partition)
+
+    self._docker_root_directory = os.path.join(
+        self._mount_path, 'var', 'lib', 'docker')
```
[the remark this answers is not in the document]
Comment (rgayon) [reply turbinia/evidence.py]: This is pretty much universal. If the docker directory is not there, then docker was started with specific options (which may or may not have been stored in /etc/docker/daemon.json)

Made this a class attribute

File: turbinia/workers/docker.py (hunk 3 above)
Comment (rgayon) [reply turbinia/workers/docker.py]: Fixed

File: turbinia/workers/docker.py (hunk 5 above)
Comment (rgayon) [reply turbinia/workers/docker.py]: Since I have to manually mount the underlying partition of parent RawDisk image (which preprocessor doesn't do), I have to clean up afterwards.

**Comment by rgayon** [comment]
sorry for procrastinating the PR for so long :(

File: turbinia/workers/docker.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (aarontp) [reply turbinia/workers/docker.py]: Yeah if there is a packaged module, let's make Turbinia depend on that.  Otherwise if it's a system dependency it is related to this issue:
https://github.com/google/turbinia/issues/312

File: turbinia/evidence.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (aarontp) [reply turbinia/evidence.py]: I'm ok with that, but can we still add a comment so that its type and description are documented?  Thanks!

File: turbinia/evidence.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (aarontp) [reply turbinia/evidence.py]: Interesting.  Ok, SG.

**Comment by aarontp** [comment]
I added a couple comments, but I'll do a more thorough review tomorrow (pending oncall load :).

File: turbinia/workers/docker.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rgayon) [reply turbinia/workers/docker.py]: SGTM.
I'm in the process of cleaning up the code of DE to make using the module a bit easier, will then rework this part.

File: turbinia/workers/docker.py (hunk 3 above)
Comment (aarontp) [reply turbinia/workers/docker.py]: Ok, maybe at some point in the future we can refactor out the actual execution code in `execute()` from the bits that update the result, etc.

File: turbinia/evidence.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (aarontp) [reply turbinia/evidence.py]: Yeah, I think that's a good idea just as a catchall.  Some of this code is starting to get a bit complex, so it would be nice to have an explicit error just in case something changes down the line.

File: turbinia/workers/docker.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+      for container_info in containers_info:
+        container_id = container_info.get('container_id')
+        found_containers.append(container_id)
+        container_evidence = DockerContainer(container_id=container_id)
+        result.add_evidence(container_evidence, evidence.config)
+      success = True
+    except TurbiniaException as e:
+      status_report = 'Error enumerating Docker containers: {0!s}'.format(e)
+
+    status_report = 'Found {0!s} containers: {1}'.format(
```
Comment (aarontp) [inline turbinia/workers/docker.py]: If there is an error in the except block above, this will overwrite the changes to status_report.  Maybe check to see if `status_report` is not empty first here?  Or maybe checking if `success == False` would be better?

File: turbinia/workers/docker.py [hunk 14]
```diff
@@ -47,30 +47,37 @@ def GetContainers(self, evidence):
...
+      raise TurbiniaException('Could not find docker-explorer script: de.py')
+    de_binary = de_paths[0]
     docker_explorer_command = [
-        'sudo', '/usr/local/bin/de.py', '-r', docker_dir, 'list',
-        'all_containers'
+        'sudo', de_binary, '-r', docker_dir, 'list', 'all_containers'
     ]
     try:
-      log.info('Running {0:s}'.format(' '.join(docker_explorer_command)))
+      log.info('Running {0}'.format(' '.join(docker_explorer_command)))
```
Comment (aarontp) [inline turbinia/workers/docker.py]: Why did you change `{0:s}` to `{0}` here?

File: turbinia/workers/docker.py [hunk 15]
```diff
@@ -47,30 +47,37 @@ def GetContainers(self, evidence):
...
       json_string = subprocess.check_output(docker_explorer_command)
     except Exception as e:
       mount_local.PostprocessUnmountPath(mount_path)
       raise TurbiniaException(
-          'Failed to run {0:s} {1!s}'.format(
+          'Failed to run {0} {1!s}'.format(
```
Comment (aarontp) [inline turbinia/workers/docker.py]: Same format string question as above.

File: turbinia/evidence.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (aarontp) [reply turbinia/evidence.py]: Thinking about this again, I really think we should try to put this into the RawDisk preprocessor, but if it can't mount any partitions, then we can just emit a warning and for those tasks that require a mounted partition we can have a check for that.  I think the longer term solution might be to have a separate evidence type for RawDiskPartition, and those get enumerated out by a separate task.  For now we can keep this as a TODO here since the other processors do similar things (e.g. GoogleCloudDisk).  I think we should move the mount that happens in the task run code though, so I'll comment on that below.

File: turbinia/workers/docker.py (hunk 5 above)
Comment (aarontp) [reply turbinia/workers/docker.py]: This same thing is going to happen with the GoogleCloudDisk AFAICT.  I think we should actually move this into the RawDisk preprocessor (and the GoogleCloudDisk preprocessor), and have a property of the disk based evidence types that is something like `mounted` that the preprocessors can set when they successfully mount the partition(s).  Then Tasks that care whether the disks are mounted are not can check for this and bail out if there is a problem, but for those Tasks that don't care (like Plaso), they can just ignore this.   As mentioned above, at a later time we can refactor this into two different evidence types (RawDisk and RawDiskPartition or maybe RawDiskVolume).

Another step for later will be disentangling the preprocessors for the other evidence types (e.g. GoogleCloudDisk) so that they either set and use the `parent_evidence` or use inheritance so that there is not code duplication for each one.

File: turbinia/evidence.py [hunk 16]
```diff
@@ -369,8 +375,60 @@ class FilteredTextFile(TextFile):
...
+    if issubclass(type(self.parent_evidence), RawDisk):
+      # Mounting the filesystem on the disk, as this is not done by the RawDisk
+      # evidence preprocessor
+      # TODO: consider making the RawDisk preprocessor always mount the
+      # partition
+      self._mount_path = mount_local.PreprocessMountDisk(
+          self.parent_evidence.loopdevice_path,
+          self.parent_evidence.mount_partition)
+
+    self._docker_root_directory = os.path.join(
```
Comment (aarontp) [inline turbinia/evidence.py]: So, right now for GoogleCloudDisk the `local_path` refers to the block device and not the mounted dir (and that gets assigned to `_mount_path` above), so I don't think this will work for that evidence type as a parent. :(.  I think as part of moving the mount calls into the RawDisk/GoogleCloudDisk preprocessors (see below comment for more), we should maybe disambiguate these into `device_path` and the already existing `mount_path` (from RawDisk), and for now we can just use `mount_path` of the parent here instead.

Long term, I'm not sure if we should keep `local_path` then or not.  We could keep it for Tasks that know how to process both (e.g. Plaso), or we could just remove it altogether and force the Task writers to explicitly use one or the other.  A quick grep shows 92 references to local_path across 20 files, so this won't be trivial though (so definitely a separate PR).

File: turbinia/processors/docker.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+def PreprocessMountDockerFS(docker_dir, container_id):
+  """Mounts a Docker container Filesystem locally.
+
+  Args:
+    docker_dir: the root Docker directory, as string.
+    container_id: the complete ID of the container, as string.
+
+  Raises:
+    TurbiniaException: if there was an error trying to mount the filesystem.
```
Comment (aarontp) [inline turbinia/processors/docker.py]: Can you add a `Returns:` entry here too?

File: turbinia/processors/docker.py [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+from turbinia import TurbiniaException
+
+log = logging.getLogger('turbinia')
+
+
+def PreprocessMountDockerFS(docker_dir, container_id):
+  """Mounts a Docker container Filesystem locally.
+
+  Args:
+    docker_dir: the root Docker directory, as string.
```
Comment (aarontp) [inline turbinia/processors/docker.py]: Tiny nit:  We can also add the type info here as `docker_dir(str)`.   I know not all of the code does that yet, but eventually I'd like it to be compliant.  More info here in the Plaso/l2t style guide:
https://github.com/log2timeline/l2tdocs/blob/master/process/Style-guide.md#docstrings

File: turbinia/workers/docker.py [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+from turbinia.processors import mount_local
+from turbinia.workers import TurbiniaTask
+
+log = logging.getLogger('turbinia')
+
+
+class DockerContainersEnumerationTask(TurbiniaTask):
+  """Enumerates Docker containers on Linux"""
+
+  def GetContainers(self, evidence):
```
Comment (aarontp) [inline turbinia/workers/docker.py]: Any chance we could get tests at least for this module? :).    The processors would be really nice to test, but those might be hard to test so e2e tests might be the only good option for that.  If you have test disks you're using for this, try to keep them around so we can add them to e2e tests later.

File: turbinia/workers/docker.py (hunk 2 above)
Comment (aarontp) [reply turbinia/workers/docker.py]: Looks like this path reference still needs to reference DEFAULT_DOCKER_DIRECTORY_PATH?

File: turbinia/workers/docker.py [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+      raise TurbiniaException('Could not find docker-explorer script: de.py')
+    de_binary = de_paths[0]
+    docker_explorer_command = [
+        'sudo', de_binary, '-r', docker_dir, 'list', 'all_containers'
+    ]
+    try:
+      log.info('Running {0}'.format(' '.join(docker_explorer_command)))
+      json_string = subprocess.check_output(docker_explorer_command)
+    except Exception as e:
+      mount_local.PostprocessUnmountPath(mount_path)
```
Comment (aarontp) [inline turbinia/workers/docker.py]: Maybe add a debug log message here indicating that the list failed so we're trying to unmount the path?  If the list failed, there's a decent chance that the unmount will fail too, so it would be nice to have something in the logs to understand what happened.

Hopefully we can actually pull out these mount/unmount from the task and then let the normal post-processors handle this.

File: turbinia/workers/docker.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rgayon) [reply turbinia/workers/docker.py]: This is done

File: turbinia/evidence.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rgayon) [reply turbinia/evidence.py]: done

File: turbinia/evidence.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (rgayon) [reply turbinia/evidence.py]: done

File: turbinia/workers/docker.py (hunk 2 above)
Comment (rgayon) [reply turbinia/workers/docker.py]: evidence here is a RawDisk. 
I feel like the DEFAULT DOCKER path should actually be part of the docker explorer lib constant or something. I'll add a TODO & filed https://github.com/google/docker-explorer/issues/80

File: turbinia/workers/docker.py (hunk 5 above)
Comment (rgayon) [reply turbinia/workers/docker.py]: moved mounting/unmounting in rawdisk

File: turbinia/workers/docker.py (hunk 13 above)
Comment (rgayon) [reply turbinia/workers/docker.py]: done

File: turbinia/evidence.py (hunk 16 above)
Comment (rgayon) [reply turbinia/evidence.py]: I think "local_path" should be the only thing that's 'visible' to tasks trying to access the data pointed to by the Evidence, and all the other paths should just be hidden & used by the pre & post processors for the management of the Evidence setup/teardown

File: turbinia/processors/docker.py (hunk 17 above)
Comment (rgayon) [reply turbinia/processors/docker.py]: done

File: turbinia/processors/docker.py (hunk 18 above)
Comment (rgayon) [reply turbinia/processors/docker.py]: done

File: turbinia/evidence.py (hunk 16 above)
Comment (rgayon) [reply turbinia/evidence.py]: Okay, the "mountening" thing needs to be its own PR :(
See https://github.com/google/turbinia/pull/385

File: turbinia/workers/docker.py (hunk 20 above)
Comment (rgayon) [reply turbinia/workers/docker.py]: this is not relevant anymore seeing the other changes

File: turbinia/workers/docker.py (hunk 19 above)
Comment (rgayon) [reply turbinia/workers/docker.py]: I would have to mock an Evidence with a mounted Docker filesystem. ANd I'd end up testing the docker_explorer module =)

File: turbinia/evidence.py [hunk 21]
```diff
@@ -132,7 +136,12 @@ def serialize(self):
     """Return JSON serializable object."""
     serialized_evidence = self.__dict__
     if self.parent_evidence:
-      serialized_evidence['parent_evidence'] = self.parent_evidence.serialize()
+      if isinstance(self.parent_evidence, dict):
+        # The parent evidence may have been serialized already
+        serialized_evidence['parent_evidence'] = self.parent_evidence
+      else:
+        serialized_evidence['parent_evidence'] = self.parent_evidence.serialize(
```
Comment (aarontp) [inline turbinia/evidence.py]: Heh, is this really the way YAPF formats this line?  Maybe start the right-hand side of the assignment with a paren so the rest of the right-hand side goes on the second line?

File: turbinia/evidence.py [hunk 22]
```diff
@@ -282,12 +304,31 @@ def __init__(self, project=None, zone=None, disk_name=None, *args, **kwargs):
     super(GoogleCloudDisk, self).__init__(*args, **kwargs)
     self.cloud_only = True
 
+    self._attached_path = None
+
   def _preprocess(self):
-    self.local_path = google_cloud.PreprocessAttachDisk(self.disk_name)
+    self._attached_path = google_cloud.PreprocessAttachDisk(self.disk_name)
```
Comment (aarontp) [inline turbinia/evidence.py]: What do you think about using just `device_path` instead of `_attached_path`?

File: turbinia/evidence.py [hunk 23]
```diff
@@ -282,12 +304,31 @@ def __init__(self, project=None, zone=None, disk_name=None, *args, **kwargs):
...
+    self.loopdevice_path = mount_local.PreprocessLosetup(self._attached_path)
```
Comment (aarontp) [inline turbinia/evidence.py]: And maybe for this one we could use volume_path (or partition_path) instead of loopdevice_path?  As per our earlier conversation I'm just trying to find common attribute names across evidence types.

Comment (aarontp) [inline turbinia/evidence.py]: Could we add a TODO(aarontp): Remove PreprocessLosetup() once https://github.com/google/turbinia/issues/78 is fixed.

File: turbinia/evidence.py [hunk 24]
```diff
@@ -282,12 +304,31 @@ def __init__(self, project=None, zone=None, disk_name=None, *args, **kwargs):
...
+          '{2!s}'.format(self.mount_partition, self.disk_name, e))
 
   def _postprocess(self):
-    google_cloud.PostprocessDetachDisk(self.disk_name, self.local_path)
-    self.local_path = None
+    if self.is_mounted:
+      mount_local.PostprocessUnmountPath(self.mount_path)
+      self.is_mounted = False
+    mount_local.PostprocessDeleteLosetup(self.loopdevice_path)
+    google_cloud.PostprocessDetachDisk(self.disk_name, self._attached_path)
```
Comment (aarontp) [inline turbinia/evidence.py]: Should we still set `self.local_path = None` here?

File: turbinia/evidence.py [hunk 25]
```diff
@@ -300,19 +341,37 @@ class GoogleCloudDiskRawEmbedded(GoogleCloudDisk):
...
         self.loopdevice_path, self.mount_partition)
-    self.local_path = os.path.join(self.mount_path, self.embedded_path)
+
+    # If the mounting of the embedded image step fails, at least local_path
+    # still points to the image.
+    self.local_path = os.path.join(
+        self._clouddisk_mount_path, self.embedded_path)
+
+    try:
+      # We now try to mount the image pointed by self.embedded_path
```
Comment (aarontp) [inline turbinia/evidence.py]: Do we also need to do a preprocessLosetup() on the inner disk image?  That way we will have a device path that PreprocessMountDisk() can turn into a valid volume device path for mounting?

File: turbinia/evidence.py [hunk 26]
```diff
@@ -300,19 +341,37 @@ class GoogleCloudDiskRawEmbedded(GoogleCloudDisk):
...
+      self.mount_path = mount_local.PreprocessMountDisk(
+          self.embedded_path, self.mount_partition)
```
Comment (aarontp) [inline turbinia/evidence.py]: The embedded_path here is the relative path inside the outer cloud disk.  I think what we want here is what is assigned to self.local_path at this point (The joining of the mount path of the outer cloud disk and the inner relative embedded_path).

File: turbinia/evidence.py [hunk 27]
```diff
@@ -300,19 +341,37 @@ class GoogleCloudDiskRawEmbedded(GoogleCloudDisk):
...
+      self.local_path = self.mount_path
+      self.is_mounted = True
+    except TurbiniaException as e:
+      log.error(
+          'Could not mount partition {0:d} for GoogleCloudDiskRawEmbedded '
+          '"{1:s}": {2!s}'.format(self.mount_partition, self.disk_name, e))
```
Comment (aarontp) [inline turbinia/evidence.py]: This error message is going to be somewhat confusing because the `mount_partition` is almost always going to be None in this preprocessor method (unless the user sets the flag), and it's only in the preprocessor itself that it will default to using `1`.  Maybe we should remove the partition here, and then make sure that the processors are including the full device path in the error message when they raise the exception?

File: turbinia/evidence.py [hunk 28]
```diff
@@ -374,3 +433,46 @@ def __init__(self, artifact_name=None):
...
+
+    self.context_dependent = True
+
+  def _preprocess(self):
+    if not self.parent_evidence:
+      raise TurbiniaException(
+          'Evidence of type DockerContainer should have a parent_evidence set.')
+    if not self.parent_evidence.is_mounted:
+      raise TurbiniaException(
+          'Parent Evidence ({0}) does not have a mounted filesystem'.format(
```
Comment (aarontp) [inline turbinia/evidence.py]: {0:s} ? :)

File: turbinia/evidence.py [hunk 29]
```diff
@@ -374,3 +433,46 @@ def __init__(self, artifact_name=None):
...
+              self.parent_evidence))
```
Comment (aarontp) [inline turbinia/evidence.py]: maybe self.parent_evidence.name?  Otherwise you can do {0!s} above and I think it will work.

File: turbinia/evidence_test.py [hunk 30]
```diff
@@ -28,15 +28,13 @@ class TestTurbiniaEvidence(unittest.TestCase):
 
   def testEvidenceSerialization(self):
     """Test that evidence serializes/unserializes."""
-    rawdisk = evidence.RawDisk(
-        name='My Evidence', local_path='/tmp/foo', mount_path='/mnt/foo')
+    rawdisk = evidence.RawDisk(name='My Evidence', local_path='/tmp/foo')
```
Comment (aarontp) [inline turbinia/evidence_test.py]: Why did the `mount_path` get removed from here?

File: turbinia/processors/docker.py [hunk 31]
```diff
@@ -0,0 +1,90 @@
...
+
+def PreprocessMountDockerFS(docker_dir, container_id):
+  """Mounts a Docker container Filesystem locally.
+
+  We use subprocess to run the DockerExplorer script, instead of using the
+  Python module, because we need to make sure all DockerExplorer code runs
+  as root.
+
+  Args:
+    docker_dir(str): the root Docker directory.
```
Comment (aarontp) [inline turbinia/processors/docker.py]: Maybe `docker_root_path` to match the evidence?

File: turbinia/processors/docker.py [hunk 32]
```diff
@@ -0,0 +1,90 @@
...
+              mount_prefix, e))
+
+  container_mount_path = tempfile.mkdtemp(prefix='turbinia', dir=mount_prefix)
+
+  log.info(
+      'Using docker_explorer to mount container {0:s} on {1:s}'.format(
+          container_id, container_mount_path))
+  # TODO(aarontp): Remove hard-coded sudo in commands:
+  # https://github.com/google/turbinia/issues/73
+  de_paths = [
```
Comment (aarontp) [inline turbinia/processors/docker.py]: This will break if this is installed in a virtual environment.    Maybe try distutils.spawn.find_executable() here too?  Not sure if that's the best solution, or how portable it is, etc, but I found that after a quick search.

File: turbinia/processors/docker.py [hunk 33]
```diff
@@ -0,0 +1,90 @@
...
+  de_binary = de_paths[0]
+  mount_cmd = [
+      'sudo', de_binary, '-r', docker_dir, 'mount', container_id,
+      container_mount_path
+  ]
+  log.info('Running: {0:s}'.format(' '.join(mount_cmd)))
+
+  try:
+    subprocess.check_call(mount_cmd)
+  except Exception as e:
```
Comment (aarontp) [inline turbinia/processors/docker.py]: Is there some less broad exception we can trap here?  Maybe subprocess.CalledProcessError?

File: turbinia/workers/docker.py [hunk 34]
```diff
@@ -0,0 +1,122 @@
...
+      evidence (Evidence): the input Evidence.
+
+    Returns:
+      a list(dict) containing information about the containers found.
+
+    Raises:
+      TurbiniaException: when the docker-explorer tool failed to run.
+    """
+
+    docker_dir = os.path.join(evidence.mount_path, 'var', 'lib', 'docker')
```
Comment (aarontp) [inline turbinia/workers/docker.py]: Didn't we add a constant for this somewhere?

File: turbinia/workers/docker.py [hunk 35]
```diff
@@ -0,0 +1,122 @@
...
+    except Exception as e:
+      raise TurbiniaException(
+          'Failed to run {0:s}: {1!s}'.format(
+              ' '.join(docker_explorer_command), e))
+
+    try:
+      containers_info = json.loads(json_string)
+    except ValueError as e:
+      raise TurbiniaException(
+          'Could not parse output of {0} : {1!s} .'.format(
```
Comment (aarontp) [inline turbinia/workers/docker.py]: {0:s}

File: turbinia/workers/docker.py [hunk 36]
```diff
@@ -0,0 +1,122 @@
...
+          found_containers.append(container_id)
+          container_evidence = DockerContainer(container_id=container_id)
+          result.add_evidence(container_evidence, evidence.config)
+        success = True
+        status_report = 'Found {0!s} containers: {1:s}'.format(
+            len(found_containers), ' '.join(found_containers))
+      except TurbiniaException as e:
+        status_report = 'Error enumerating Docker containers: {0!s}'.format(e)
+    else:
+      status_report = (
```
Comment (aarontp) [inline turbinia/workers/docker.py]: You're setting the `status_report` for the negative case here twice (once here and once at the top of the if statement).  Maybe just set it once before the `if` statement, and then you can drop the `else`?

File: turbinia/workers/docker.py [hunk 37]
```diff
@@ -0,0 +1,122 @@
...
+          'Evidence {0!s} did not expose a mounted file system'.format(
```
Comment (aarontp) [inline turbinia/workers/docker.py]: BTW, I like the similar error message above better, it's a bit more clear.

File: turbinia/evidence.py (hunk 24 above)
Comment (rgayon) [reply turbinia/evidence.py]: So turns out setting self.local_path to None is a bad idea, because IIUC one Evidence object can be postprocessed, but later on re-preprocessed. If we unset local_path, we can't re-preprocess

File: turbinia/evidence.py (hunk 22 above)
Comment (rgayon) [reply turbinia/evidence.py]: Sure. I'm keeping the underscore in front of attributes names that we keep around just so we can later use them when doing postprocess

File: turbinia/evidence.py (hunk 23 above)
Comment (rgayon) [reply turbinia/evidence.py]: this is going to be /dev/loopX  and not /dev/loopXpY
the mount preprocessor takes  /dev/loopX path and a partition number to build the path to /dev/loopXpY.
A TODO could be to have evidence.GetVolumePath(volume_number) that does this

device_path =  google_cloud.PreprocessAttachDisk
volume_path = self.GetVolumePath(device_path, volume_number) // or partition_number
mount_path = mount_local.mount(volume_path)

Comment (rgayon) [reply turbinia/evidence.py]: TODOs have been sprinkled

File: turbinia/evidence.py (hunk 25 above)
Comment (rgayon) [reply turbinia/evidence.py]: indeed. FIxed

**Comment by aarontp** [comment]
@rgayon Can we drop this PR in favor of #465 then?
